### DISCUSSION TITLE
Add two cost parameters for each vehicle type

### DISCUSSION THREAD

**Discussion by tcokyasar44** [google/or-tools#2545] [Feature requests] [answered]
Just to increase the complexity, how about adding two different cost parameters for the same vehicle type? For example, a hybrid vehicle runs the first x distance/time on EV-mode, and hence the cost is lower than the distance/time after x. Meanwhile, another vehicle type with a fixed cost can be a benchmark. Is it possible to model this?

**Comment by orwant** [comment, marked as answer]
I really don't think you want to go there :-)  Instead, choose a lower
overall operating cost for the hybrid.

That keeps the modeling easy, and it also better reflects reality: your
typical hybrid car experience is not "electric for these 5 miles, then gas
for the next 5" but a blend as the car turns the engine off and on all the
time (speaking as a former Prius owner). Similarly for planes, which burn
fuel at different rates depending on altitude and speed: you *could*
attempt to model that, and even make those decision variables, but you'd be
upping the complexity of your problem so much that it changes how you'd
solve it.

Jon

**Reply by tcokyasar44** [reply]
Hey Jon,

I admit that introducing even more complexity to a readily NP-hard problem is not ideal. Can I at least model the problem for two types of vehicles where the first set of vehicle charges c_1 cost per time unit and the second set charges c_2 cost?

**Reply by paultrow** [reply]
You can define a separate cost function for each vehicle using SetArcCostEvaluatorOfVehicle (https://github.com/google/or-tools/blob/v8.1/ortools/constraint_solver/routing.h#L898).  You need to create a separate callback for each vehicle, which returns the travel costs for that vehicle, and pass the callbacks to RegisterTransitCallback - see https://developers.google.com/optimization/routing/tsp#dist_callback1.
